torch/fx/passes/fuser_utils.py

```diff
@@ -0,0 +1,228 @@
+    # assumption: nodes are already sorted in topo order
+
+    for node in nodes:
+        assert node.graph.owning_module is gm, f"{node} doesn't belong to passed in graph module {gm._get_name()}"
+        assert not node._erased, f"{node} has been removed from owning graph"
+        assert node in gm.graph.nodes, f"{node} is not found in graph module {gm._get_name()}"
+
+    # validates partition doesn't introduce dependency circles in the graph
+    assert validate_partition(nodes), "Invalid partition, found dependency cycles"
+
+    subgraph = torch.fx.Graph()
+
+    node_to_placeholder = {}  # mapping of nodes from old graph to placeholder in new graph
+    node_map = {}       # mapping of nodes from old graph to new graph
+
+    # handles inputs throught graph.node_copy's arg_transform functions
+    def remap_inputs(x):
+        if x.op == "get_attr":
+            # TODO: do we really need copy the get_attr node into the graph?
+            # do something here
+            pass
+
+        if x in nodes:
+            # x is inside subgraph, return the copied node
+            # the node should have been copied aleady, as we are copying graph in the topological order
+            return node_map[x]
+
+        if x not in node_to_placeholder:
+            # x is not in subgraph, create a new placeholder for subgraph
+            node_to_placeholder[x] = subgraph.placeholder(x.name, type_expr=x.type)
+
+        return node_to_placeholder[x]
+
+    # copy nodes in topological order
+    for node in nodes:
+        new_node = subgraph.node_copy(node, remap_inputs)
+        node_map[node] = new_node
+
+    # handles outputs
+    output_mapping = {}  # mapping from old output to new outputs
+
+    for node in nodes:
+        for user_node in node.users:
+            if user_node not in nodes:
+                # external user node, need to expose as an output
+                output_mapping[node] = node_map[node]
+
+    # outs contain nodes in the new subgraph
+    original_outputs = tuple(output_mapping.keys())
+    outs = tuple(output_mapping.values())
+
+    # Take care of the args of FX output node. If there's a single
+    # output then the output node args is like (output_single), else
+    # if there're multiple outputs then the output node args is like
+    # ((output_0, output_1, ...)).
+    subgraph.output(outs[0] if len(outs) == 1 else outs)
+
+    # lint to ensure correctness
+    subgraph.lint()
+
+    submodule = copy_module_attributes(gm, subgraph)
+
+    sub_gm = torch.fx.GraphModule(submodule, subgraph, class_name=partition_name)
+
+    # TODO: fix this
+    sub_gm.name = partition_name
+    sub_gm.orig_inputs = tuple(node_to_placeholder.keys())
+    sub_gm.orig_outputs = original_outputs
+
+    return sub_gm
+
+def insert_subgm(gm, sub_gm, original_nodes):
+    # assign sub_gm into gm
+    setattr(gm, sub_gm.name, sub_gm)
+
+    # Create a call_module node in main graph.
+    module_node = gm.graph.call_module(
```

Comment:
should we set the insertion point?

Reply:
we call legalize_graph(gm) after the insertion..., which performs topological sort to make the graph valid... 
so we don't need to explicitly specify insertion point

Reply:
but let me try setting insertion point anyway..

Reply:
I think setting the insertion point would have problem with horizontal fusion.
Consider following graph with nodes 1~8. Original node's order could be 1,2,3,4,5,6,7,8. 
Assuming the fusion group is [2,3,6,7]. In this case, a valid node order would be 1, 5, fused, 4, 8. 
Simply controlling where to insert fused node is not sufficient to guarantee a valid topological order. 

![image](https://user-images.githubusercontent.com/9906745/174919360-226ff8d7-e58d-4ad1-8924-6b8789b36cf8.png)